api/src/main/java/io/grpc/MetricInstrumentRegistry.java

```diff
@@ -50,7 +62,7 @@ public static synchronized MetricInstrumentRegistry getDefaultRegistry() {
    * Returns a list of registered metric instruments.
    */
   public List<MetricInstrument> getMetricInstruments() {
-    return Collections.unmodifiableList(metricInstruments);
+    return ImmutableList.copyOf(metricInstruments);
```

Comment:
This can throw `ConcurrentModificationException`, because it isn't synchronized with adds. Since this is called rarely, I suggest we throw a lock around this and call it good enough for now. If we want to optimize it, we could save this copy and clear it when adding, but let's not bother for the moment.

Because of this copy, manually resizing the list is unnecessary and doesn't actually do much; the list will do that for you. To get the benefit of the `registerLongGauge()`, you'd want to remove the copy here. To do that you'd either:
1. Treat the ArrayList like an array. You'd `add(null)` entries up to its capacity when creating it. Then to actually add an entry you'd use `set()
2. Use an array, and then Arrays.asList() to produce a return value

In both cases you'd use `Collections.unmodifiableList(list.subList(0, instrumentIndexAlloc))` for the return value here.

Since `getMetricInstruments()` is called rarely, it's probably expedient to accept the copy and avoid manually managing the list size (delete registerLongGauge()). We can then come back to it.

Reply:
Thanks for explaining in detail about potential issue. Updated to use array, and added a lock. PTAL